packages/devtools_app/lib/src/screens/performance/rebuild_counts.dart

```diff
@@ -20,112 +21,297 @@ import '../../primitives/utils.dart';
+      final events = <int>[];
+      for (RebuildLocation rebuild in rebuilds) {
+        events
+          ..add(rebuild.location.id)
+          ..add(rebuild.buildCount);
       }
+      frames.add({'frameNumber': id, 'events': events});
+    });
+    return <String, dynamic>{
+      'locations': locationMap.toJson(),
+      'frames': frames,
+    };
+  }
+
+  void processRebuildEvent(Map<String, dynamic> json) {
+    // parse locations
+    if (json.containsKey('locations')) {
+      locationMap.processLocationMap(json['locations'] as Map<String, dynamic>);
     }
 
+    processRebuildsForFrame(json);
+
+    // We've updated the build counts and possibly the locations.
+    _stats.dataChanged();
+  }
+
+  void clearFromRestart() {
+    _rebuildsForFrame.clear();
+    locationMap.clear();
+    _stats.clear();
+  }
+
+  void clearAllCounts() {
+    _stats.clear();
+    _rebuildsForFrame.clear();
+  }
+
+  void processRebuildsForFrame(Map<String, dynamic> json) {
+    final int frameNumber = json['frameNumber'];
     // parse events
     final List<int> events = (json['events'] as List).cast<int>();
+    final rebuildsForFrame = <RebuildLocation>[];
     for (int i = 0; i < events.length; i += 2) {
       final id = events[i];
       final count = events[i + 1];
 
-      final location = _locationMap[id];
-      location?.buildCount += count;
+      final location = locationMap[id];
+      rebuildsForFrame.add(RebuildLocation(location, buildCount: count));
+      _stats.incrementCount(location, count);
+    }
+    _rebuildsForFrame[frameNumber] = rebuildsForFrame;
+    while (_rebuildsForFrame.length > rebuildFrameCacheSize) {
+      _rebuildsForFrame.remove(_rebuildsForFrame.keys.first);
     }
-
-    // We've updated the build counts and possibly the locations.
-    _locations.notifyListeners();
   }
+}
 
-  void clearFromReload() {
-    // TODO(devoncarew): We need to call this when we see a hot reload or restart.
+class Location {
+  Location({required this.id, this.path, this.line, this.column, this.name});
 
-    _locationMap.clear();
-    _locations.clear();
-  }
+  int id;
 
-  void clearCurrentCounts() {
-    for (final location in _locationMap.values) {
-      location.buildCount = 0;
-    }
+  /// Either all of path, line, column, and name are null or none are.
+  String? path;
+  int? line;
+  int? column;
+  String? name;
 
-    // We've updated the build counts.
-    _locations.notifyListeners();
-  }
+  bool get isResolved => path != null;
 }
 
 class RebuildLocation {
-  RebuildLocation({this.id, this.path, this.line, this.column, this.name});
+  RebuildLocation(this.location, {this.buildCount = 0});
+  Location location;
+  int buildCount;
+}
 
-  final int? id;
-  final String? path;
-  final int? line;
-  final int? column;
-  final String? name;
+/// Helper class to merge together rebuild metrics from multiple sources.
+/// For example, rebuild metrics on a per frame basis and total metrics since
```

Comment:
nit: new line after first sentence of dart doc

Reply:
Done